scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/Utils.scala

```diff
@@ -0,0 +1,198 @@
+      case NoFile                                            => "There is not at least one file in this metadata file"
+      case MissingParent(missingParentId)                    => s"Parent id $missingParentId does not exist in the json file"
+
+    }
+
+    case class ParentTypeInvalid(parentTypeOpt: Option[Type], fileTypeOpt: Option[Type], id: UUID) extends ErrorMessage
+
+    case class CircularDependency(id: UUID, existingParents: List[UUID]) extends ErrorMessage
+
+    case class NoAssetChildren(id: UUID) extends ErrorMessage
+
+    case object NoTopLevelFolder extends ErrorMessage
+
+    case object NoAsset extends ErrorMessage
+
+    case object NoFile extends ErrorMessage
+
+    case class MissingParent(missingParentId: UUID) extends ErrorMessage
+
+    given Encoder[ErrorMessage] = Encoder.encodeString.contramap[ErrorMessage](_.show)
+    given Encoder[SingleValidationResult] = deriveEncoder[SingleValidationResult]
+    given Encoder[WholeFileValidationResult] = deriveEncoder[WholeFileValidationResult]
+
+    case class SingleValidationError(results: List[SingleValidationResult]) extends Throwable
+    case class SingleValidationResult(json: Json, errors: List[String])
+    case class WholeFileValidationResult(errors: List[String], singleResults: List[SingleValidationResult]) {
+      def anyErrors: Boolean = errors.nonEmpty || singleResults.flatMap(_.errors).nonEmpty
+    }
+  end ErrorMessage
+
+  sealed trait MandatoryFields:
+    def id: UUID
+
+    def parentId: Option[UUID]
+
+    def getType: Type = this match
+      case _: MandatoryArchiveFolderFields => ArchiveFolder
+      case _: MandatoryContentFolderFields => ContentFolder
+      case _: MandatoryAssetFields         => Asset
+      case _: MandatoryFileFields          => File
+  end MandatoryFields
+
+  object MandatoryFields:
+
+    given Configuration = Configuration.default.withDefaults
+
+    case class ObjectCounts(assetCount: Int, fileCount: Int, topLevelCount: Int)
+
+    given Decoder[RepresentationType] = (c: HCursor) => c.as[String].map(RepresentationType.valueOf)
+
+    given Decoder[IdField] = deriveDecoder[IdField]
+
+    case class MandatoryArchiveFolderFields(id: UUID, name: String, parentId: Option[UUID], series: Option[String]) extends MandatoryFields
+
+    case class MandatoryContentFolderFields(id: UUID, name: String, parentId: Option[UUID], series: Option[String]) extends MandatoryFields
+
+    case class MandatoryAssetFields(
+        id: UUID,
+        parentId: Option[UUID],
+        series: Option[String],
+        digitalAssetSource: String,
+        originalFiles: List[UUID],
+        originalMetadataFiles: List[UUID],
+        transferCompleteDatetime: OffsetDateTime,
+        transferringBody: String,
+        upstreamSystem: String
+    ) extends MandatoryFields
+
+    case class MandatoryFileFields(
+        id: UUID,
+        name: String,
+        parentId: Option[UUID],
```

Comment:
Technically yes but the trait needs to have the parent ID on it as an Option otherwise it would make everything much more complicated. This is a case of accepting a minor semantic error in order to make the code much easier to read.